### ISSUE TITLE
F# server doesn't persist failed fn/handler executions

### ISSUE THREAD

**Issue by StachuDotNet** [darklang/dark#3620]
(WIP - still adjusting the wording of this PR, and researching)
Title likely needs some adjustment.

This applies only to the F# backend - the issue stems from a difference between the F# and OCaml backends, but we believe that the F# backend is doing the right thing, and that the client should handle this better. This issue has been extracted from a conversation in #3610

--- 
Issue:
Create a handler that will result in an error, and play it.
We'd expect the new error to pop up now, but instead we see `This trace returns <Incomplete>` and `Click Play to execute function`

https://user-images.githubusercontent.com/906686/159572072-d41d6399-c822-4db4-ae6f-9ffd894d48ff.mp4

--- 
Context:
When I execute this function, an `execute_function` API call is made. The response includes both the result, as well as any `touched_tlids`:
```
{
    "result": [
        "DError",
        [
            [ "SourceId", 2077505382, 714461266 ],
            "expecting str"
        ]
    ],
    "hash": "7MvOlhz7MMr-Ce95_UJdRlst14TX8WHcijuk5Hr5O5Pr4zx1mAF1HRkEaLx4t9h6",
    "hashVersion": 1,
    "touched_tlids": [
        2077505382
    ],
    "unlocked_dbs": [ 820101588, 1315303241, 1541193719, 1852027420, 2124904283 ]
}
```

The F# server makes a point of always including the relevant TLID in the `touched_tlids` bag, while the OCaml server in some cases doesn't include it.

If the `touched_tlids` contains anything, subsequent requests are made to the API, to route `get_trace_data`.

Once that data is loaded, the _previous_ execution seems to paint over the quickly-shown error message, with the messaging we display in the video.

**Comment by StachuDotNet** [comment]
New discovery: if I execute the REPL even once against the OCaml server, I can then see the error message in the F#-loaded client. (though, it's finnicky, and I cannot produce _new_ errors - only load the OCaml-produced ones)

**Comment by StachuDotNet** [comment]
Small update here, though I think I'm close to finishing the picture:

When `execute_function` results in a failure,
- the OCaml version successfully stores the results in the DB (`function_results_v3`, `function_arguments`)
- the F# version does not store those results in the DB

So, this is a backend issue, I believe. Currently tracing where the F# server is failing to persist these. Once the backend is properly storing this, the load of traces `get_trace_data` would result in the right error being displayed, and we shouldn't need to adjust anything in the client code.

**Comment by StachuDotNet** [comment]
Found the code that failed. Considering next step

https://user-images.githubusercontent.com/906686/159782315-435514af-3298-409e-be3e-202f33af6d69.mp4

.

**Comment by StachuDotNet** [comment]
@pbiggar My intuition here is that we should fix this in the backend, ensuring that such failures are stored. Let me know if that's going down the wrong path. Reading starting here (https://github.com/darklang/dark/issues/3620#issuecomment-1076705717) is enough context.

**Comment by pbiggar** [comment]
Nice find! Yeah, the backend is definitely the place to fix that!

**Comment by pbiggar** [comment]
There's a hacky attempt in https://github.com/darklang/dark/pull/3637 and I'm going to attempt a non-hacky version.

**Comment by StachuDotNet** [comment]
Summary of findings so far:

The issue occurs in `Interpreter.fs`, namely in the fn `execFn` [here](https://github.com/darklang/dark/blob/2592d0a6b0af8e07b5a31953e688f66df0d8353f/fsharp-backend/src/LibExecution/Interpreter.fs#L589).
Once we've gathered the relevant state, have collected the args, we go to actually evaluate the fn.v

There's a `try` / `with` expression that looks roughly like this:
```
try
  match fn.fn with
  | StdLib f ->
    // let result = ...

    if (conditions) then
      do! state.tracing.storeFnResult fnRecord arglist result

    return result
  | PackageFunction (_tlid, body) ->
    // same as above
  | UserFunction (tlid, body) ->
    // same above
with
| ExceptionType1 -> ...
| ExceptionType2 -> ...
| ExceptionType3 -> ...
| ExceptionType4 -> ...
| ...
```

Depending on type (stdlib, package, user) of function, whether the function is marked as Previewable, and whether we encounter errors while evaluating, we may or may not _intend_ to call `state.tracing.storeFnResult`. This specific GitHub Issue is that exceptions raised prevent `storeFnResult` from ever being called.

This requires some sort of refactor to ensure we always call `storeFnResult` in the cases that we expect to.

A few ideas of how to refactor towards this:
1. (the hack in #3637) add a `mutable` within the code to store whether we intend to call the function; update the `try` branch to adjust this as appropriate, and update the `with` logic to ensure it's always called in the failure cases
2. compute `shouldStoreFnResult` separately from the current `try` branch (i.e. slightly above it), replicating the logic currently in there; surround the try/with with a `let result = ...` binding; call `storeFnResult` after the try/with is done
3. add `try`/`with` expressions around individual branches under the _main_ `try`. (i.e. wrap `let! result = eval state argsWithGlobals body` with a try/with); handle the failure cases individually there.

If we go down the road of option 2, and simplify some logic, I eventually get to:
```
let shouldStoreFnResults =
  match fn.fn with
  | StdLib f ->
    fn.previewable <> Pure && state.tracing.realOrPreview <> Preview
  | PackageFunction (_tlid, _body) ->
    match TypeChecker.checkFunctionCall Map.empty fn args with
    | Ok () ->
      let firstCheck =
        match (state.tracing.realOrPreview, state.tracing.loadFnResult fnRecord arglist)
        with
        | Preview, Some _ -> false
        | Preview, None when fn.previewable <> Pure -> false
        | _ -> true

      firstCheck ||  fn.previewable <> Pure
    | Error errs -> false
  | UserFunction (tlid, body) ->
    match TypeChecker.checkFunctionCall state.program.userTypes fn args with
    | Ok () ->
      match (state.tracing.realOrPreview, state.tracing.loadFnResult fnRecord arglist)
      with
      | Preview, Some _ -> false
      | _ -> true
    | Error errs -> false
```

which (while can be simplified further) has a few problems.
First, some of this logic is pretty opaque without context of the main `try` logic. Though, maybe some of the original logic is flawed.
Second, now that we have that binding, what do we do now? My natural reaction would be to put that 'above' the try/with, to wrap the try/with with a `let result = ...` binding like option 3 suggests, but, there are cases where we intend to store 1 thing as the result, but _return_ some variant of that. (see [here](https://github.com/darklang/dark/blob/main/fsharp-backend/src/LibExecution/Interpreter.fs#L697-L704)) So, with this, we'd be storing something other than what we intended. That said, [here](https://github.com/StachuDotNet/dark/commit/040bf2e56887588b045293ba6cc12a3157e65c1a)'s a branch with that attempt.

The idea of teasing out the "should we store the results?" logic from the rest sounds appealing, and I think we could simplify the above a bit, but I'm not sure how to handle the current logic occasionally stores one dval, and returns another.

I've only just considered option 3, haven't yet tested deeper to see if that's realistic.

**Comment by pbiggar** [comment]
Looking into this now. In addition to the trace problem, I think sending the extra tlid (the current handler) in touched_tlids causes an extra call to get_trace_data.

**Comment by StachuDotNet** [comment]
> Looking into this now. In addition to the trace problem, I think sending the extra tlid (the current handler) in touched_tlids causes an extra call to get_trace_data.

It does, but once a fix is in place for the storage of errors (tested), that `get_trace_data` results in expected UI, at least in the cases I saw - I figured that would be "ok" at worst. Maybe even advantageous?

**Comment by pbiggar** [comment]
So in ast.ml, we don't actually catch exceptions for PackageFunction and UserFunctions. I guess we could go back to that.

**Comment by pbiggar** [comment]
It seems there are some edge cases here we didn't support, which should be tested:

- [ ] are there tests for exceptions/errors in calls to package manager functions?
  - [ ] simple wrong number of arguments
  - [ ] with/without unwrapFromErrorRails
- [ ] are there tests for exceptions/errors in calls to user functions functions?
  - [ ] simple wrong number of arguments
  - [ ] with/without unwrapFromErrorRails

**Comment by pbiggar** [comment]
Fixed in #3638, followups in #3639.
